File: python/step2_eval.py
import sys, traceback
import readline
import mal_types as types
import reader, printer

# read
def READ(str):
    return reader.read_str(str)

# eval
def eval_ast(ast, env):
    if type(ast) == types.Symbol:
        try:    return env[ast]
        except: raise Exception("'" + ast + "' not found")
    elif type(ast) == list:
        return list(map(lambda a: EVAL(a, env), ast))
    else:
        return ast

def EVAL(ast, env):
    if not type(ast) == list:
        return eval_ast(ast, env)

    # apply list
    if len(ast) == 0: return ast
    el = eval_ast(ast, env)
    f = el[0]
    return f(*el[1:])


# print
def PRINT(exp):
    return printer.pr_str(exp)

# repl
repl_env = {'+': lambda a,b: a+b,
            '-': lambda a,b: a-b,
            '*': lambda a,b: a*b,
            '/': lambda a,b: int(a/b)}

def REP(str):
    return PRINT(EVAL(READ(str), repl_env))

# repl loop
while True:
    try:
        line = input("user> ")
        if line == None: break
        if line == "": continue
        print(REP(line))
    except reader.Blank: continue
    except EOFError: break
    except Exception as e:
        print("".join(traceback.format_exception(*sys.exc_info())))

File: ruby/step2_eval.rb
require "readline"
require_relative "reader"
require_relative "printer"

# read
def READ(str)
    return read_str(str)
end

# eval
def eval_ast(ast, env)
    return case ast
        when Symbol
            raise "'" + ast.to_s + "' not found" if not env.key? ast
            env[ast]
        when Array
            Array.new ast.map{|a| EVAL(a, env)}
        else
            ast
    end
end

def EVAL(ast, env)
    if not ast.is_a? Array
        return eval_ast(ast, env)
    end
    if ast.empty?
        return ast
    end

    # apply list
    el = eval_ast(ast, env)
    f = el[0]
    return f[*el.drop(1)]
end

# print
def PRINT(exp)
    return pr_str(exp)
end

# repl
repl_env = {:+ => lambda {|a,b| a + b},
            :- => lambda {|a,b| a - b},
            :* => lambda {|a,b| a * b},
            :/ => lambda {|a,b| a / b}}
def REP(str, env)
    return PRINT(EVAL(READ(str), env))
end

# repl loop
while line = Readline.readline("user> ", true)
    begin
        puts REP(line, repl_env)
    rescue Exception => e
        puts "Error: #{e}"
        puts "\t#{e.backtrace.join("\n\t")}"
    end
end
